cfg/builder/builder_finalize.cc

```diff
@@ -333,50 +341,52 @@ void CFGBuilder::fillInBlockArguments(core::Context ctx, const CFG::ReadsAndWrit
     // This solution is  (|BB| + |symbols-mentioned|) * (|cycles|) + |answer_size| in complexity.
     // making this quadratic in anything will be bad.
 
-    const vector<UnorderedSet<core::LocalVariable>> &readsByBlock = RnW.reads;
-    const vector<UnorderedSet<core::LocalVariable>> &writesByBlock = RnW.writes;
-    const vector<UnorderedSet<core::LocalVariable>> &deadByBlock = RnW.dead;
+    const vector<vector<int>> &readsByBlock = RnW.reads;
+    const vector<vector<int>> &writesByBlock = RnW.writes;
+    const vector<vector<int>> &deadByBlock = RnW.dead;
 
-    // iterate ver basic blocks in reverse and found upper bounds on what could a block need.
-    vector<UnorderedSet<core::LocalVariable>> upperBounds1(cfg.maxBasicBlockId);
+    // iterate over basic blocks in reverse and found upper bounds on what could a block need.
+    vector<vector<int>> upperBounds1;
     bool changed = true;
     {
         Timer timeit(ctx.state.tracer(), "upperBounds1");
-        for (BasicBlock *bb : cfg.forwardsTopoSort) {
-            auto &upperBoundsForBlock = upperBounds1[bb->id];
-            upperBoundsForBlock.insert(readsByBlock[bb->id].begin(), readsByBlock[bb->id].end());
-        }
+        upperBounds1 = readsByBlock;
         while (changed) {
             changed = false;
             for (BasicBlock *bb : cfg.forwardsTopoSort) {
                 auto &upperBoundsForBlock = upperBounds1[bb->id];
-                int sz = upperBoundsForBlock.size();
+                const auto sz = upperBoundsForBlock.size();
                 if (bb->bexit.thenb != cfg.deadBlock()) {
-                    upperBoundsForBlock.insert(upperBounds1[bb->bexit.thenb->id].begin(),
-                                               upperBounds1[bb->bexit.thenb->id].end());
+                    setUnionInplace(upperBoundsForBlock, upperBounds1[bb->bexit.thenb->id]);
                 }
                 if (bb->bexit.elseb != cfg.deadBlock()) {
-                    upperBoundsForBlock.insert(upperBounds1[bb->bexit.elseb->id].begin(),
-                                               upperBounds1[bb->bexit.elseb->id].end());
+                    setUnionInplace(upperBoundsForBlock, upperBounds1[bb->bexit.elseb->id]);
                 }
+
                 // Any variable that we write and do not read is dead on entry to
                 // this block, and we do not require it.
-                for (auto dead : deadByBlock[bb->id]) {
-                    // TODO(nelhage) We can't erase for variables inside loops, due
-                    // to how our "pinning" type inference works. We can remove this
-                    // inner condition when we get a better type inference
-                    // algorithm.
-                    if (bb->outerLoops <= cfg.minLoops[dead]) {
-                        upperBoundsForBlock.erase(dead);
+                const auto &deadForBlock = deadByBlock[bb->id];
+                if (!deadForBlock.empty()) {
```

Comment:
Avoids a vector allocation and a call to `setDifferenceInplace` if nothing is dead.